text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1006 @@
+
+Thus we have a marker (`const`) to add in front of functions that requires the function body to only contain things const eval can handle. This in turn allows a `const` annotated function to be called from const contexts, as you now have a guarantee it will stay callable.
+
+When calling a trait method, this simple scheme (that works great for free functions and inherent methods) does not work.
+
+Throughout this document, we'll be revisiting the example below. Method syntax and `dyn Trait` problems all also exist with static method calls, so we'll stick with the latter to have the simplest examples possible.
+
+```rust
+const fn default<T: Default>() -> T {
+    T::default()
+}
+
+// Could also be `const fn`, but that's an orthogonal change
+fn compile_time_default<T: Default>() -> T {
+    const { T::default() }
+}
+```
+
+Neither of the above should (or do) compile.
+The first, because you could pass any type T whose default impl could
+
+* mutate a global static,
+* read from a file, or
+* just allocate memory,
+
+which are all not possible right now in const code, and some can't be done in Rust in const code at all.
+
+It should be possible to write `default` in a way that allows it to be called in const contexts
+for types whose `Default` impl's `default` method satisfies all rules that `const fn` must satisfy
+(including some annotation that guarantees this won't break by accident).
+It must always be possible to call `default` outside of const contexts with no limitations on the generic parameters that may be passed.
+
+Similarly it should be possible to write `compile_time_default` in a way that also requires calls
+outside of const contexts to only pass generic parameters whose `Default::default` method satisifies
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait impls
+
+It is now allowed to prefix a trait name in an impl block with `const`, marking that this `impl`'s type is now allowed to
+have methods of this `impl`'s trait to be called in const contexts (if all where bounds hold, like usual, but more on this later).
+
+An example looks as follows:
+
+```rust
+impl const Trait for Type {}
+```
+
+Such impls require that the trait is a `const trait`.
+
+All method bodies in a const trait impl are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+### Const traits
+
+Traits need to opt-in to being allowed to have const trait impls. Thus you need to declare your traits by prefixing the `trait` keyword with `const`:
+
+```rust
+const trait Trait {}
```

Comment:
One thing which is worth briefly touching upon is the idea of traits *without* methods being marked `const`, since it technically does nothing *by itself*. It's a useful API guarantee if you choose to later add methods, but in the future, we may have traits which have an API guarantee to never have methods (e.g. `marker_trait`), and it's not clear how we should handle those. Even though it's speculating about unstable features, I think it's worth at least mentioning.

Reply:
I don't know what to write here. A trait without methods being const can be useful as a marker in blanket impls or just constraining what types can go into a function because you want to limit it without there being a type system reason in the body of the function (some unsafe stuff maybe?).